Add everything [draft, to be closed]

Virviil opened the pull request:
This PR adds bindings to models and decoders.

Looking forward fast review, i want to continue implementing features based on this pr.
Thank you!

seanmor5 reviewed (commented):
Hi @Virviil the Elixir looks good to me, but I do not know Rust. @philss are you able to take a look at that?

What are your plans with these additional features? How are you using these?

philss commented:
> @philss are you able to take a look at that?

@seanmor5 sure! I'm going to review it soon.

philss commented on lib/tokenizers/model/bpe.ex:
```diff
@@ -0,0 +1,50 @@
+defmodule Tokenizers.Model.BPE do
+  @typedoc """
+  Options for model initialisation. All values are optional.
```
I would change the docs here to say that "all options can be omitted". Because saying that they are optional can imply that they can be `nil`. WDYT?

Also, maybe we can validate the accepted options using [`Keyword.validate!/2`](https://hexdocs.pm/elixir/Keyword.html#validate!/2). WDYT?

Virviil replied:
1. Y, mb it will be better. I'll change the entire codebase.
2. Values are validated by the mechanism of `NifTaggedEnum`. As keyword is basically a list of tuples2, each element can be parsed as `NifTaggedEnum`, while the keyword itself is a `Vec<T: NifTaggedEnum>`. Nif will rise with `:bad_arg`, mb it's not `:ok|:err` tuple, but `Keyword.validate!` will rise in the same way.

philss commented on native/ex_tokenizers/src/decoders.rs:
```diff
@@ -0,0 +1,233 @@
+#[rustler::nif]
+fn decoders_info(
+    decoder: ExTokenizersDecoder,
+) -> Result<HashMap<String, DetailValue>, ExTokenizersError> {
```
I think we may want to optimize this result to return a Term representing the Map. It's not needed for this PR, but just to keep an eye on it.

Virviil replied:
I haven't got it. Can you please share a short snippet mb?

philss commented on native/ex_tokenizers/src/models.rs:
```diff
@@ -0,0 +1,349 @@
+    Ok(match &model.resource.0 {
+        ModelWrapper::BPE(model) => HashMap::from([
+            (
+                String::from("model_type"),
```
In the `decoders_info`, you used the `"model_type".into()` syntax to create a String. What about change it here - in this entire file - to follow that pattern?

Virviil replied:
Good catch

philss commented on native/ex_tokenizers/src/util.rs:
```diff
@@ -0,0 +1,11 @@
+use rustler::NifUntaggedEnum;
+
+#[derive(NifUntaggedEnum)]
+pub enum DetailValue {
```
If we change that `HashMap` for an encoded Map as `Term`, we can get rid of this enum, since all the primitives can be encoded already.

philss reviewed (commented):
Looks good to me!

josevalim commented on lib/tokenizers/decoder/byte_fallback.ex:
```diff
@@ -0,0 +1,7 @@
+defmodule Tokenizers.Decoder.ByteFallback do
+    @doc """
```
```suggestion
  @doc """
```

josevalim commented on lib/tokenizers/decoder/bpe.ex:
```diff
@@ -0,0 +1,7 @@
+  Creates new BPE decoder
+  """
+  @spec new(suffix :: String.t()) :: {:ok, Tokenizers.Decoder.t()} | {:error, any()}
+  defdelegate new(suffix \\ "</w>"), to: Tokenizers.Native, as: :decoders_bpe
```
I am not very happy with one module per decoder. Could we instead have `Tokenizers.Decoder.bpe`, `Tokenizers.Decoder.byte_fallback`, etc? The smaller abstraction in Elixir is a function and one function modules are frequently an anti-pattern.

josevalim replied:
And yes, I agree Models should be modules :)

josevalim commented on test/model/bpe_test.exs:
```diff
@@ -0,0 +1,52 @@
+  use ExUnit.Case, async: true
+  doctest Tokenizers.Model.BPE
+
+  describe "Initialized from memory" do
```
Nitpick: make sure all `describe` and `test` start in lowercase to be consistent with the rest of the project. :)

Virviil commented:
@josevalim @seanmor5 @philss 

So I've realized that all the modules are basically coupled with each other, so I decided to put full implementation in just a single PR. Hope it will be possible to review it.

* I've updated info mechanism with macros - so it's more handy now
* I've removed `Result` return in favor of just raising in many reasonable places
* Is far as i understand - external API didn't change at all, only grew with new functions. You can check tests, and livemd to ensure.
* Now API is fully compatible with _official_ ones
* I've added Quicktour from hugging face site converted to livemd, so it's possible to check training capabilities
* Some entities are _mutable_ underneese, it's by intention to keep common API with other versions. Probably this part is mostly discussible.

josevalim commented on lib/tokenizers/model.ex:
```diff
@@ -3,30 +3,39 @@ defmodule Tokenizers.Model do
+  """
+  @spec save(model :: t(), folder :: String.t(), prefix :: String.t()) ::
+          {:ok, file_pathes :: [String.t()]} | {:error, any()}
+  defdelegate save(model, folder, prefix \\ ""), to: Tokenizers.Native, as: :models_save
```
What does the prefix option do? Can we either remove it or make it an option?

josevalim commented on test/tokenizers/pre_tokenizer_test.exs:
```diff
@@ -0,0 +1,93 @@
+defmodule Tokenizers.PreTokenizerTest do
```
These tests seem to be wrong. :)

josevalim commented:
Hi @Virviil, this looks great and very comprehensive but there are a few problems:

1. The PR is now very large :)

2. This PR removes APIs which are important to Elixir/Nx, such as `get_u32_*` functions. For efficiency, sometimes it is better to get a u32 binary in Elixir rather than a list of numbers (especially if we want to convert it to a tensor)

3. It also has breaking changes to tokenizers

4. Some functionality is not being tested, such as pre-tokenizers

Finally, note I prefer default arguments to be options since it gives us more space for "growth" in the future without introducing breaking changes.

However, given the PR is large, it is hard currently to see all of the places we could improve. Therefore, could you please break this PR into smaller ones, here is a suggestion:

1. One PR for models + decoding, as originally proposed
2. One PR for normalizers, pre and post processing
3. One PR for changing the tokenizers/encoding APIs but please avoid breaking changes (unless strictly necessary, feel free to drop comments if there is something you believe it should be removed)
4. API for training / added tokens

Thank you!

josevalim commented:
Hi @Virviil, this looks great and very comprehensive but there are a few problems:

1. The PR is now very large :)

2. This PR removes APIs which are important to Elixir/Nx, such as `get_u32_*` functions. For efficiency, sometimes it is better to get a u32 binary in Elixir rather than a list of numbers (especially if we want to convert it to a tensor)

3. It also has breaking changes to tokenizers

4. Some functionality is not being tested, such as pre-tokenizers

Finally, note I prefer default arguments to be options since it gives us more space for "growth" in the future without introducing breaking changes.

However, given the PR is large, it is hard currently to see all of the places we could improve. Therefore, could you please break this PR into smaller ones, here is a suggestion:

1. One PR for models + decoding, as originally proposed
2. One PR for normalizers and pre-processing
3. One PR for changing the tokenizers/encoding APIs but please avoid breaking changes (unless strictly necessary, feel free to drop comments if there is something you believe it should be removed)
4. Then follow up PRs for training, added tokens, post-processing...

Thank you!
